File: Source/Game/Podium/TeamPodium.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;
using WCSharp.Shared.Data;

public static class TeamPodium
{
    private static Queue<(player Player, Point position)> PodiumQueue = new Queue<(player Player, Point position)>();
    private static List<unit> MovedUnits = new List<unit>();
    private static string PodiumType = "";
    private static string Color = Colors.COLOR_YELLOW_ORANGE;
    private const string MVP = "MVP";
    private const string MostSaves = "most saves";

    public static void BeginPodiumActions()
    {
        PodiumUtil.SetCameraToPodium();
        Utility.SimpleTimer(3.0f, ProcessPodiumTypeActions);
    }

    private static void EnqueueMVPPlayer()
    {
        var topRatios = PodiumUtil.SortPlayersByHighestRatio();
        var podiumPositions = PodiumManager.PodiumSpots;
        for (int i = topRatios.Count - 1; i >= 0; i--)
        {
            var player = topRatios[i];
            var position = podiumPositions[i];
            PodiumQueue.Enqueue((player, position));
        }
        PodiumType = MVP;
    }

    private static void EnqueueTopSavesPlayer()
    {
        var topSaves = PodiumUtil.SortPlayersBySaves();
        var podiumPositions = PodiumManager.PodiumSpots;
        for (int i = topSaves.Count - 1; i >= 0; i--)
        {
            var player = topSaves[i];
            var position = podiumPositions[i];
            PodiumQueue.Enqueue((player, position));
        }
        PodiumType = MostSaves;
    }

    private static void ProcessNextPodiumAction()
    {
        if (PodiumQueue.Count == 0)
        {
            ProcessPodiumTypeActions();
            return;
        }
        var (player, position) = PodiumQueue.Dequeue();
        var kitty = Globals.ALL_KITTIES[player].Unit;
        kitty.SetPosition(position.X, position.Y);
        kitty.SetFacing(270);
        kitty.IsPaused = true;
        MovedUnits.Add(kitty);
        Console.WriteLine($"{Colors.PlayerNameColored(player)}{Color} earned {PodiumUtil.PlacementString(PodiumQueue.Count + 1)} place for {PodiumType} with {GetStatBasedOnType(player)}|r");
        Utility.SimpleTimer(5.0f, ProcessNextPodiumAction);
    }

    /// <summary>
    /// Topscore => Saves => Ratio => Streak .. End
    /// </summary>
    private static void ProcessPodiumTypeActions()
    {
        PodiumQueue.Clear();
        PodiumUtil.ClearPodiumUnits(MovedUnits);
        switch (PodiumType)
        {
            case "":
                EnqueueMVPPlayer();
                break;

            case MVP:
                EnqueueTopSavesPlayer();
                break;

            case MostSaves:
                PodiumUtil.EndingGameThankyou();
                return;

            default:
                break;
        }
        ProcessNextPodiumAction();
    }

    private static string GetStatBasedOnType(player player)
    {
        var stats = Globals.ALL_KITTIES[player].CurrentStats;
        switch (PodiumType)
        {
            case MostSaves:
                return stats.TotalSaves.ToString();

            case MVP:
                return stats.TotalDeaths == 0 ? stats.TotalSaves.ToString("F2") + " ratio." : ((float)stats.TotalSaves / stats.TotalDeaths).ToString("F2") + " ratio";

            default:
                return "n/a";
        }
    }
}

File: src/Game/Podium/TeamPodium.ts
import { Globals } from 'src/Global/Globals'
import { Colors } from 'src/Utility/Colors/Colors'
import { ColorUtils } from 'src/Utility/Colors/ColorUtils'
import { Queue } from 'src/Utility/Queue'
import { Utility } from 'src/Utility/Utility'
import { MapPlayer, Point, Unit } from 'w3ts'
import { PodiumUtil } from './PodiumUtil'

export class TeamPodium {
    private static PodiumQueue = new Queue<[MapPlayer, Point]>()
    private static MovedUnits: Unit[] = []
    private static PodiumType: string = ''
    private static Color: string = Colors.COLOR_YELLOW_ORANGE
    private static MVP: string = 'MVP'
    private static MostSaves: string = 'saves: most'

    public static BeginPodiumActions() {
        PodiumUtil.SetCameraToPodium()
        Utility.SimpleTimer(3.0, TeamPodium.ProcessPodiumTypeActions)
    }

    private static EnqueueMVPPlayer() {
        let topRatios = PodiumUtil.SortPlayersByHighestRatio()
        let podiumPositions = PodiumUtil.PodiumSpots
        for (let i: number = topRatios.length - 1; i >= 0; i--) {
            let player = topRatios[i]
            let position = podiumPositions[i]
            TeamPodium.PodiumQueue.enqueue([player, position])
        }
        TeamPodium.PodiumType = TeamPodium.MVP
    }

    private static EnqueueTopSavesPlayer() {
        let topSaves = PodiumUtil.SortPlayersBySaves()
        let podiumPositions = PodiumUtil.PodiumSpots
        for (let i: number = topSaves.length - 1; i >= 0; i--) {
            let player = topSaves[i]
            let position = podiumPositions[i]
            TeamPodium.PodiumQueue.enqueue([player, position])
        }
        TeamPodium.PodiumType = TeamPodium.MostSaves
    }

    private static ProcessNextPodiumAction() {
        if (TeamPodium.PodiumQueue.length === 0) {
            TeamPodium.ProcessPodiumTypeActions()
            return
        }
        let [player, position] = TeamPodium.PodiumQueue.dequeue()!
        let kitty = Globals.ALL_KITTIES.get(player)!.Unit
        kitty.setPosition(position.x, position.y)
        kitty.setFacingEx(270)
        kitty.paused = true
        TeamPodium.MovedUnits.push(kitty)
        print(
            `${ColorUtils.PlayerNameColored(player)}${TeamPodium.Color} earned ${PodiumUtil.PlacementString(TeamPodium.PodiumQueue.length + 1)} for: place ${TeamPodium.PodiumType} with ${TeamPodium.GetStatBasedOnType(player)}|r`
        )
        Utility.SimpleTimer(5.0, TeamPodium.ProcessNextPodiumAction)
    }

    /// <summary>
    /// Topscore => Saves => Ratio => Streak .. End
    /// </summary>
    private static ProcessPodiumTypeActions() {
        TeamPodium.PodiumQueue.length = 0
        PodiumUtil.ClearPodiumUnits(TeamPodium.MovedUnits)
        switch (TeamPodium.PodiumType) {
            case '':
                TeamPodium.EnqueueMVPPlayer()
                break

            case TeamPodium.MVP:
                TeamPodium.EnqueueTopSavesPlayer()
                break

            case TeamPodium.MostSaves:
                PodiumUtil.EndingGameThankyou()
                return

            default:
                break
        }
        TeamPodium.ProcessNextPodiumAction()
    }

    private static GetStatBasedOnType(player: MapPlayer) {
        let stats = Globals.ALL_KITTIES.get(player)!.CurrentStats
        switch (TeamPodium.PodiumType) {
            case TeamPodium.MostSaves:
                return stats.TotalSaves.toString()

            case TeamPodium.MVP:
                return stats.TotalDeaths === 0
                    ? stats.TotalSaves.toFixed(2) + ' ratio.'
                    : (stats.TotalSaves / stats.TotalDeaths).toFixed(2) + ' ratio'

            default:
                return 'n/a'
        }
    }
}
